File: examples/javascript/aws_account_id_extractor.js
function base32Decode(input) {
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  let buffer = 0, bitsLeft = 0, output = [];

  for (let char of input) {
    const val = alphabet.indexOf(char.toUpperCase());
    if (val === -1) continue; // Ignore invalid characters
    buffer = (buffer << 5) | val;
    bitsLeft += 5;

    if (bitsLeft >= 8) {
      bitsLeft -= 8;
      output.push((buffer >> bitsLeft) & 0xFF);
      buffer &= (1 << bitsLeft) - 1;
    }
  }
  return Uint8Array.from(output);
}

function AWSAccountFromAWSKeyID(AWSKeyID) {
  if (AWSKeyID.length <= 4) return "Invalid Key ID";
  let trimmed = AWSKeyID.slice(4);
  let decoded = base32Decode(trimmed);
  let bigIntValue = BigInt('0x' + Array.from(decoded.slice(0, 6)).map(b => b.toString(16).padStart(2, '0')).join(''));
  let mask = BigInt('0x7fffffffff80');
  return ((bigIntValue & mask) >> 7n).toString().padStart(12, '0');
}

// Example Usage
let awsKeyID = "ABCDABCDEFGHIJKLMNOPQRSTUVWX";
console.log("AWS Account ID:", AWSAccountFromAWSKeyID(awsKeyID));

File: examples/java/AWSAccountIDExtractor.java
import java.math.BigInteger;
import java.util.Base64;

public class AWSAccountIDExtractor {

    private static final String ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

    public static byte[] base32Decode(String input) {
        int buffer = 0, bitsLeft = 0;
        byte[] output = new byte[(int) Math.ceil(input.length() * 5 / 8)];
        int index = 0;

        for (char c : input.toCharArray()) {
            int value = ALPHABET.indexOf(Character.toUpperCase(c));
            if (value == -1) continue; // Ignore invalid characters
            buffer = (buffer << 5) | value;
            bitsLeft += 5;

            if (bitsLeft >= 8) {
                bitsLeft -= 8;
                output[index++] = (byte) (buffer >> bitsLeft);
                buffer &= (1 << bitsLeft) - 1;
            }
        }
        return java.util.Arrays.copyOf(output, index);
    }

    public static String getAWSAccountID(String awsKeyID) {
        if (awsKeyID.length() <= 4) return "Invalid Key ID";
        String trimmed = awsKeyID.substring(4);
        byte[] decoded = base32Decode(trimmed);
        byte[] accountBytes = java.util.Arrays.copyOfRange(decoded, 0, 6);
        BigInteger accountID = new BigInteger(1, accountBytes).shiftRight(7);
        return String.format("%012d", accountID);
    }

    public static void main(String[] args) {
        String awsKey = "ABCDABCDEFGHIJKLMNOPQRSTUVWX";
        System.out.println("AWS Account ID: " + getAWSAccountID(awsKey));
    }
}
